File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/api/models/mailbox/v2/BookletDetailDto.kt
@file:Suppress("PackageNaming")

package app.pantopus.android.data.api.models.mailbox.v2

import app.pantopus.android.data.api.models.common.JsonValue

/**
 * Booklet sub-payload decoded from `mail.object_payload` when
 * `mail_type == "booklet"`. Backend stores this as untyped JSON in S3.
 * [decodeFromObjectPayload] returns null when the payload doesn't list
 * any pages.
 */
data class BookletDetailDto(
    val pages: List<String>,
    val summary: String?,
    val pageCount: Int,
) {
    companion object {
        fun decodeFromObjectPayload(payload: JsonValue?): BookletDetailDto? {
            if (payload == null) return null
            val rawPages = payload["pages"] as? List<*> ?: emptyList<Any?>()
            val pages = rawPages.mapNotNull { it as? String }
            if (pages.isEmpty()) return null
            val pageCount =
                (payload["page_count"] as? Number)?.toInt() ?: pages.size
            return BookletDetailDto(
                pages = pages,
                summary = payload["summary"] as? String,
                pageCount = pageCount,
            )
        }
    }
}

File: frontend/apps/ios/Pantopus/Core/Networking/Models/Mailbox/V2/BookletDetailDTO.swift
//
//  BookletDetailDTO.swift
//  Pantopus
//
//  Booklet sub-payload decoded from `mail.object_payload` when
//  `mail_type == "booklet"`. Backend stores this as untyped JSON in S3
//  (route handler at `backend/routes/mailboxV2.js:412`); the DTO is
//  defensive and `decode(from:)` returns nil if the payload doesn't
//  carry a single page URL.
//

import Foundation

/// Booklet detail payload — drives the FrameBooklet swiper + facts.
public struct BookletDetailDTO: Sendable, Hashable {
    public let pages: [URL]
    public let summary: String?
    public let pageCount: Int

    public init(pages: [URL], summary: String?, pageCount: Int) {
        self.pages = pages
        self.summary = summary
        self.pageCount = pageCount
    }

    public static func decode(from value: JSONValue?) -> BookletDetailDTO? {
        guard let dict = value?.dictValue else { return nil }
        let pageURLs: [URL] = (dict["pages"]?.arrayValue ?? [])
            .compactMap(\.stringValue)
            .compactMap(URL.init(string:))
        guard !pageURLs.isEmpty else { return nil }
        return BookletDetailDTO(
            pages: pageURLs,
            summary: dict["summary"]?.stringValue,
            pageCount: dict["page_count"]?.numberValue.map { Int($0) } ?? pageURLs.count
        )
    }
}
